### ISSUE TITLE
Nodes seemingly garbage collected before queries

### ISSUE THREAD

**Issue by alex-statsig** [facebook/relay#4070]
When using useLazyLoadQuery, despite a query being in the store the fields no longer are. In this query, the data is not stale. Also, the query does not seem to have been evicted from the cache (useLazyLoadQuery does not suspend), even though the query is not retained. This leads to a response from useLazyLoadQuery that does not match the types, and causes @required's to fail and throw.

Note the edge still seems to show in the store, but the ref it points to is gone (which seems to be why the query things its data is present?)

To reproduce:
Load a page that will throw an error if data is not present in the query (in my case I use `@required(action: THROW)`). Then navigate to other queries to fill up the relay store with ~1000 objects in my code (but not enough to have the maximum number of queries cached or else the whole query will be removed from the store). After returning, since the query is still in the store useLazyLoadQuery does not throw suspense. However, the data required to fulfill the query is no longer in the store (in this case, "weekly_velocity_chart" was evicted).
Reproduced on relay 13 and 14

I could not find documentation describing when individual records would be evicted from the store and not the whole query. I'm wondering if this is some race condition where the query is being evicted and some items have been removed, but not marked as such yet.

Another thing of note with my setup is I do use both react-relay and relay-hooks in various parts of the codebase.

Code:
```
const query = graphql`
  query HomeVelocityCardContentQuery {
    viewer_company {
      weekly_velocity_chart
        @required(action: THROW) {
        ...RelayLineChart_chart
      }
    }
  }
`;

export default function HomeVelocityCardContent({
  selectedView,
}: Props): null | JSX.Element {
  const data = useLazyLoadQuery<HomeVelocityCardContentQuery>(
    query,
    { exposureType: selectedView },
    { fetchPolicy: 'store-and-network' },
  );
  return null;
}
```

This repros throughout our code for different routes, seemingly after a large number of queries causing the store to fill up.

**Comment by alex-statsig** [comment]
Any update on this?

**Comment by captbaritone** [comment]
Relay's GC is based on queries and records (objects and all of their fields). The expected behavior is that as long as a query is mounted, all of the records reachable by that query (with its current arguments) should be retained. Once the query becomes unmounted, any records that none of the other active queries can reach are eligible to get GCed.

There should not be any place where single records get GCed.

I suspect what's happening here is that when the `@required(action: THROW)` throws it unmounts the component (thus releasing the query). Other queries that are still mounted can reach the other records in your query, and so they remain retained. GC then kicks in and frees your `weekly_velocity_chart` record.

When you remount, the `weekly_velocity_chart` record is missing and your component needs to fetch again. It sounds like that last part is not happening for you? Instead of re-fetching it just renders with incomplete data? Given your fetch-policy, dues it _also_ make a network request and eventually fill in the missing data?

To debug you should be able to set a breakpoint before `useLazyLoadQuery` and step through. There should be a call to [`DataChecker.check`](https://github.com/facebook/relay/blob/main/packages/relay-runtime/store/DataChecker.js#L88) which checks in the store to see if we have the required data to render (or if we need to check). Does that get called? If so, what does it return, and what happens next? If not, can you figure out why not?

To debug GC issues you can pass a logger to Relay store. You should then see events telling you about which records are getting GCed: https://github.com/facebook/relay/blob/main/packages/relay-runtime/store/RelayModernStore.js#L631

**Comment by alex-statsig** [comment]
Thanks, I'll try out this debugging flow. I don't think the @required is unmounting it, but perhaps suspense is? I'm not perfectly familiar with suspense under the hood, but I think it would unmount and remount

**Comment by captbaritone** [comment]
Yeah, when a component suspends it is unmounted and replaced in the tree with the fallback ui.

**Comment by alex-statsig** [comment]
@captbaritone I think I figured out what was happening here, I'm not sure if its expected behavior; I will try and validate this further with a better repro, but here's a sample flow:

Sample schema:
```graphql
type Query {
  viewer: Viewer
}
type Viewer {
  non_nullable_field: Int!
}
```

1) Component A fires a query:
```graphql
query ComponentA {
  viewer {
    non_nullable_field
  }
}
```
This might receive a response like `viewer: { non_nullable_field: 1 }`

2) Component B fires a query, checking if the user is logged in
```graphql
query ComponentB {
  viewer {
    __typename
  }
}
```
This could receive a response like `viewer: null` if the viewer's session expirees. This will cause ComponentA to re-render with `viewer=null`, a legitimate environment to render with (ComponentA must handle `viewer` being null since its nullable in the schema)

3) Component B (or perhaps another component) re-queries, and receives data such as `viewer: { __typename: "Viewer" }`

This now causes ComponentA to re-render again with a non-null `viewer` field, but no data for the `non_nullable_field`. This is not a valid environment to render ComponentA with, and will likely lead to a runtime error due to the typescript types saying that `non_nullable_field` must be non-null (if `viewer` is non-null).

**Comment by captbaritone** [comment]
Ah, yes! This is a known issue. See: https://relay.dev/docs/debugging/why-null/#graph-relationship-change

We recently (v18) added @throwOnFieldError which turns this case into an explicit runtime error (to be caught by a React error boundary hopefully).

https://relay.dev/docs/guides/throw-on-field-error-directive/

**Comment by alex-statsig** [comment]
> Ah, yes! This is a known issue. See: https://relay.dev/docs/debugging/why-null/#graph-relationship-change

Thanks, I now have a vague memory of reading a discussion on this situation (and the tradeoffs of consistency + up-to-dateness + avoiding-refetching that relay makes, which leads to this bug).


> 
> We recently (v18) added @throwOnFieldError which turns this case into an explicit runtime error (to be caught by a React error boundary hopefully).
> 
> https://relay.dev/docs/guides/throw-on-field-error-directive/

Ah, I was curious if the new directives could help, that's good to know. That should at least make it easier to identify / consistently handle these cases (rather than silently getting nulls); I'll need to think more about how to fix the root issue still, so I can ideally avoid hitting an error boundary for this.

Would I need to add `@throwOnFieldError` to all fragments, or would it bubble up to the query if I just added it there?
